// radixdlt/build/main/modules/wallet/RadixKeyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixUtil_1 = tslib_1.__importDefault(require("../common/RadixUtil"));
var RadixAtomModel_1 = require("../RadixAtomModel");
var RadixUniverse_1 = require("../universe/RadixUniverse");
var elliptic_1 = tslib_1.__importDefault(require("elliptic"));
var bs58_1 = tslib_1.__importDefault(require("bs58"));
var universe = RadixUniverse_1.radixUniverse;
var ec = new elliptic_1.default.ec('secp256k1');
var RadixKeyPair = /** @class */ (function () {
    function RadixKeyPair() {
    }
    RadixKeyPair.generateNew = function () {
        var radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.genKeyPair();
        return radixKeyPair;
    };
    RadixKeyPair.fromAddress = function (address) {
        var raw = Array.prototype.slice.call(bs58_1.default.decode(address), 0);
        // Universe check
        if (universe.getMagicByte() != raw[0]) {
            throw new Error('Address is from a different universe');
        }
        // Checksum
        var check = RadixUtil_1.default.hash(raw.splice(0, raw.length - 4), 0, raw.length - 4);
        for (var i = 0; i < 4; i++) {
            if (check[i] != raw[raw.length - 4 + i]) {
                throw new Error('Invalid address');
            }
        }
        raw = Array.prototype.slice.call(bs58_1.default.decode(address), 0);
        var radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPublic(raw.splice(1, raw.length - 5));
        return radixKeyPair;
    };
    RadixKeyPair.fromPublic = function (publicKey) {
        if (!publicKey) {
            throw 'Missing public key';
        }
        if (publicKey.length != 33) {
            throw 'Public key must be 33 bytes, but was ' + publicKey.length;
        }
        var radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPublic(publicKey);
        return radixKeyPair;
    };
    RadixKeyPair.fromPrivate = function (privateKey) {
        var radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPrivate(privateKey);
        return radixKeyPair;
    };
    RadixKeyPair.fromRadixECKeyPair = function (keyPair) {
        return this.fromPublic(keyPair.public.data);
    };
    RadixKeyPair.prototype.getAddress = function () {
        var publicKey = this.keyPair.getPublic().encode('be', true);
        var addressBytes = [];
        addressBytes[0] = universe.getMagicByte();
        for (var i = 0; i < publicKey.length; i++) {
            addressBytes[i + 1] = publicKey[i];
        }
        var check = RadixUtil_1.default.hash(addressBytes, 0, publicKey.length + 1);
        for (var i = 0; i < 4; i++) {
            addressBytes[publicKey.length + 1 + i] = check[i];
        }
        return bs58_1.default.encode(Buffer.from(addressBytes));
    };
    RadixKeyPair.prototype.getHash = function () {
        return RadixUtil_1.default.hash(this.getPublic(), 0, this.getPublic().length);
    };
    RadixKeyPair.prototype.getUID = function () {
        var hash = this.getHash();
        return new RadixAtomModel_1.RadixEUID(hash.slice(0, 12));
    };
    RadixKeyPair.prototype.getShard = function () {
        return RadixUtil_1.default.longFromBigInt(this.getUID().value);
    };
    RadixKeyPair.prototype.getPublic = function () {
        return Buffer.from(this.keyPair.getPublic().encode('be', true));
    };
    RadixKeyPair.prototype.getPrivate = function (enc) {
        return this.keyPair.getPrivate(enc);
    };
    RadixKeyPair.prototype.sign = function (data) {
        var signature = this.keyPair.sign(data);
        return RadixAtomModel_1.RadixSignature.fromEllasticSignature(signature);
    };
    RadixKeyPair.prototype.verify = function (data, signature) {
        return this.keyPair.verify(data, elliptic_1.default.Signature(signature));
    };
    RadixKeyPair.prototype.equals = function (otherKeyPair) {
        return this.getAddress() == otherKeyPair.getAddress();
    };
    RadixKeyPair.prototype.toString = function () {
        return this.getAddress();
    };
    return RadixKeyPair;
}());
exports.default = RadixKeyPair;
