// radixdlt/build/main/modules/cache/RadixNEDBAtomCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var nedb_1 = tslib_1.__importDefault(require("nedb"));
var RadixAtomModel_1 = require("../RadixAtomModel");
var RadixLogger_1 = require("../common/RadixLogger");
var RadixNEDBAtomCache = /** @class */ (function () {
    /**
     * Creates an instance of radix nedbatom cache.
     * @param filename File path on disk in Node, path in localStorage in the browser
     */
    function RadixNEDBAtomCache(filename) {
        var _this = this;
        this.storeAtom = function (atom) {
            return _this.notExists({ _id: atom._id })
                .then(function () {
                // logger.info('Atom doesnt exist, storing ', atom._id, atom)
                // Add particle ids?
                // Serialize
                var serializedAtom = atom.toJson();
                serializedAtom['_id'] = atom._id;
                // logger.info(serializedAtom)
                // Store
                return _this.insert(serializedAtom);
            })
                .then(function (newDoc) {
                // Success
                // logger.info('Atom stored in DB', newDoc)
                return atom;
            })
                .catch(function (error) {
                RadixLogger_1.logger.trace('Atom already in DB');
            });
        };
        this.getAtoms = function (keyPair) {
            // Find
            var query = {};
            // Filter by destination
            if (keyPair) {
                var destination = keyPair.getUID().toJson();
                query = { destinations: destination };
            }
            // logger.info(query)
            return _this.find(query).then(function (atoms) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var deserialized, atoms_1, atoms_1_1, atom, _a, _b, e_1_1;
                var e_1, _c;
                return tslib_1.__generator(this, function (_d) {
                    switch (_d.label) {
                        case 0:
                            deserialized = [];
                            _d.label = 1;
                        case 1:
                            _d.trys.push([1, 6, 7, 8]);
                            atoms_1 = tslib_1.__values(atoms), atoms_1_1 = atoms_1.next();
                            _d.label = 2;
                        case 2:
                            if (!!atoms_1_1.done) return [3 /*break*/, 5];
                            atom = atoms_1_1.value;
                            _b = (_a = deserialized).push;
                            return [4 /*yield*/, this.asyncDeserialize(atom)];
                        case 3:
                            _b.apply(_a, [_d.sent()]);
                            _d.label = 4;
                        case 4:
                            atoms_1_1 = atoms_1.next();
                            return [3 /*break*/, 2];
                        case 5: return [3 /*break*/, 8];
                        case 6:
                            e_1_1 = _d.sent();
                            e_1 = { error: e_1_1 };
                            return [3 /*break*/, 8];
                        case 7:
                            try {
                                if (atoms_1_1 && !atoms_1_1.done && (_c = atoms_1.return)) _c.call(atoms_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                            return [7 /*endfinally*/];
                        case 8: return [2 /*return*/, deserialized];
                    }
                });
            }); });
        };
        // Promise wrappers for nedb
        this.findOne = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.findOne(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.notExists = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.findOne(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else if (!doc) {
                        resolve(true);
                    }
                    reject('Atom already in db');
                });
            });
        };
        this.find = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.find(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.insert = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.insert(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.remove = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.remove(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.db = new nedb_1.default({
            filename: filename,
            autoload: true,
        });
    }
    /**
     * Clears all atoms from the cache
     */
    RadixNEDBAtomCache.prototype.reset = function () {
        this.db.remove({}, { multi: true }, function (error, numRemoved) {
            // Do nothing
        });
    };
    RadixNEDBAtomCache.prototype.asyncDeserialize = function (atom) {
        return new Promise(function (resolve, reject) {
            setTimeout(function () {
                resolve(RadixAtomModel_1.RadixSerializer.fromJson(atom));
            }, 0);
        });
    };
    RadixNEDBAtomCache.prototype.deleteAtom = function (atom) {
        return this.remove({ _id: atom._id });
    };
    return RadixNEDBAtomCache;
}());
exports.default = RadixNEDBAtomCache;
