// radixdlt/build/module/modules/universe/RadixUniverseConfig.js
import universe_development from '../common/universe_development';
import universe_highgarden from '../common/universe_highgarden';
import universe_alphanet from '../common/universe_alphanet';
import universe_alphanet2 from '../common/universe_alphanet2';
import universe_sunstone from '../common/universe_sunstone';
import Long from 'long';
export default class RadixUniverseConfig {
    constructor(obj) {
        this.port = obj.port;
        this.name = obj.name;
        this.description = obj.description;
        this.type = obj.type;
        this.timestamp = obj.timestamp;
        this.creator = Buffer.from(obj.creator.value, 'base64');
        this.genesis = obj.genesis;
        // this.magic = this.creator.getUID().value
        //     .mul(new BN(31 * 13 * 7 * this.port * this.type))
        //     .mul(new BN(this.timestamp))
        this.magic = obj.magic;
    }
    getMagic() {
        return this.magic;
    }
    getMagicByte() {
        return Long.fromNumber(this.magic)
            .and(0xff)
            .toNumber();
    }
}
RadixUniverseConfig.WINTERFELL = new RadixUniverseConfig(universe_development);
RadixUniverseConfig.WINTERFELL_LOCAL = new RadixUniverseConfig(universe_development);
RadixUniverseConfig.SUNSTONE = new RadixUniverseConfig(universe_sunstone);
RadixUniverseConfig.HIGHGARDEN = new RadixUniverseConfig(universe_highgarden);
RadixUniverseConfig.ALPHANET = new RadixUniverseConfig(universe_alphanet);
RadixUniverseConfig.ALPHANET2 = new RadixUniverseConfig(universe_alphanet2);
