// radixdlt/build/module/modules/wallet/RadixKeyPair.js
import RadixUtil from '../common/RadixUtil';
import { RadixSignature, RadixEUID } from '../RadixAtomModel';
import { radixUniverse } from '../universe/RadixUniverse';
import EC from 'elliptic';
import bs58 from 'bs58';
const universe = radixUniverse;
const ec = new EC.ec('secp256k1');
export default class RadixKeyPair {
    static generateNew() {
        let radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.genKeyPair();
        return radixKeyPair;
    }
    static fromAddress(address) {
        let raw = Array.prototype.slice.call(bs58.decode(address), 0);
        // Universe check
        if (universe.getMagicByte() != raw[0]) {
            throw new Error('Address is from a different universe');
        }
        // Checksum
        let check = RadixUtil.hash(raw.splice(0, raw.length - 4), 0, raw.length - 4);
        for (let i = 0; i < 4; i++) {
            if (check[i] != raw[raw.length - 4 + i]) {
                throw new Error('Invalid address');
            }
        }
        raw = Array.prototype.slice.call(bs58.decode(address), 0);
        let radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPublic(raw.splice(1, raw.length - 5));
        return radixKeyPair;
    }
    static fromPublic(publicKey) {
        if (!publicKey) {
            throw 'Missing public key';
        }
        if (publicKey.length != 33) {
            throw 'Public key must be 33 bytes, but was ' + publicKey.length;
        }
        let radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPublic(publicKey);
        return radixKeyPair;
    }
    static fromPrivate(privateKey) {
        let radixKeyPair = new RadixKeyPair();
        radixKeyPair.keyPair = ec.keyFromPrivate(privateKey);
        return radixKeyPair;
    }
    static fromRadixECKeyPair(keyPair) {
        return this.fromPublic(keyPair.public.data);
    }
    getAddress() {
        let publicKey = this.keyPair.getPublic().encode('be', true);
        let addressBytes = [];
        addressBytes[0] = universe.getMagicByte();
        for (let i = 0; i < publicKey.length; i++) {
            addressBytes[i + 1] = publicKey[i];
        }
        let check = RadixUtil.hash(addressBytes, 0, publicKey.length + 1);
        for (let i = 0; i < 4; i++) {
            addressBytes[publicKey.length + 1 + i] = check[i];
        }
        return bs58.encode(Buffer.from(addressBytes));
    }
    getHash() {
        return RadixUtil.hash(this.getPublic(), 0, this.getPublic().length);
    }
    getUID() {
        let hash = this.getHash();
        return new RadixEUID(hash.slice(0, 12));
    }
    getShard() {
        return RadixUtil.longFromBigInt(this.getUID().value);
    }
    getPublic() {
        return Buffer.from(this.keyPair.getPublic().encode('be', true));
    }
    getPrivate(enc) {
        return this.keyPair.getPrivate(enc);
    }
    sign(data) {
        let signature = this.keyPair.sign(data);
        return RadixSignature.fromEllasticSignature(signature);
    }
    verify(data, signature) {
        return this.keyPair.verify(data, EC.Signature(signature));
    }
    equals(otherKeyPair) {
        return this.getAddress() == otherKeyPair.getAddress();
    }
    toString() {
        return this.getAddress();
    }
}
