// radixdlt/build/main/modules/common/RadixEUID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixUtil_1 = tslib_1.__importDefault(require("./RadixUtil"));
var RadixAtomModel_1 = require("../RadixAtomModel");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var RadixEUID = /** @class */ (function () {
    function RadixEUID(value) {
        if (Array.isArray(value)) {
            this.value = RadixUtil_1.default.bigIntFromByteArray(Buffer.from(value));
        }
        else if (Buffer.isBuffer(value)) {
            this.value = RadixUtil_1.default.bigIntFromByteArray(value);
        }
        else {
            this.value = new bn_js_1.default(value);
        }
    }
    RadixEUID.fromJson = function (data) {
        return new RadixEUID(data.value);
    };
    RadixEUID.prototype.equals = function (euid) {
        return this.value.eq(euid.value);
    };
    RadixEUID.prototype.toJson = function () {
        return {
            serializer: RadixEUID.SERIALIZER,
            value: this.value.toString()
        };
    };
    RadixEUID.prototype.toByteArray = function () {
        var type = RadixAtomModel_1.DataTypes.EUID;
        var data = RadixUtil_1.default.byteArrayFromBigInt(this.value);
        var length = data.length;
        var output = Buffer.alloc(length + 5);
        output.writeUInt8(type, 0);
        output.writeUInt32BE(length, 1);
        data.copy(output, 5);
        return output;
    };
    RadixEUID.prototype.toString = function () {
        return this.value.toString();
    };
    RadixEUID.SERIALIZER = 'EUID';
    return RadixEUID;
}());
exports.default = RadixEUID;
