// radixdlt/build/main/modules/atom/RadixBasicContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var typescript_map_1 = require("typescript-map");
var RadixUtil_1 = tslib_1.__importDefault(require("../common/RadixUtil"));
var RadixAtomModel_1 = require("../RadixAtomModel");
var RadixBasicContainer = /** @class */ (function () {
    function RadixBasicContainer(json) {
        this.version = 100;
        this.serializationProperties = new Array();
        if (json) {
            for (var key in json) {
                if (key == 'constructor' || key == 'serializationProperties') {
                    continue;
                }
                this[key] = json[key];
            }
        }
        // this.serializationProperties.push('serializer')
        this.serializationProperties.push('version');
    }
    Object.defineProperty(RadixBasicContainer.prototype, "serializer", {
        get: function () {
            return this.constructor['SERIALIZER'];
        },
        set: function (serializer) {
            // Do nothing
        },
        enumerable: true,
        configurable: true
    });
    RadixBasicContainer.prototype.toJson = function () {
        var output = { serializer: 0 };
        for (var key in this) {
            var serialized = RadixAtomModel_1.RadixSerializer.toJson(this[key]);
            if (serialized) {
                output[key] = serialized;
            }
        }
        output.serializer = this.constructor['SERIALIZER'];
        return output;
    };
    RadixBasicContainer.prototype.toByteArray = function () {
        var e_1, _a, e_2, _b;
        // Generic object
        var type = RadixAtomModel_1.DataTypes.OBJECT;
        var length = 0;
        // Serialize all properties
        // Build a map sorted by property name
        var map = new typescript_map_1.TSMap();
        try {
            for (var _c = tslib_1.__values(this.serializationProperties), _d = _c.next(); !_d.done; _d = _c.next()) {
                var key = _d.value;
                if (!(key in this)) {
                    continue;
                }
                var serializedValue = RadixAtomModel_1.RadixSerializer.toByteArray(this[key]);
                length += key.length + 1 + serializedValue.length;
                map.sortedSet(key, serializedValue);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        // Write everything to the buffer
        var output = Buffer.alloc(length + 5);
        output.writeUInt8(type, 0);
        output.writeUInt32BE(length, 1);
        var position = 5;
        try {
            for (var _e = tslib_1.__values(map.keys()), _f = _e.next(); !_f.done; _f = _e.next()) {
                var key = _f.value;
                var value = map.get(key);
                output.writeUInt8(key.length, position);
                position++;
                output.write(key, position);
                position += key.length;
                value.copy(output, position);
                position += value.length;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return output;
    };
    RadixBasicContainer.prototype.getHash = function () {
        var serialized = this.toByteArray();
        return RadixUtil_1.default.hash(serialized);
    };
    RadixBasicContainer.prototype.getHID = function () {
        var hash = this.getHash();
        return new RadixAtomModel_1.RadixEUID(hash.slice(0, 12));
    };
    Object.defineProperty(RadixBasicContainer.prototype, "hid", {
        get: function () {
            return this.getHID();
        },
        set: function (hid) {
            // Do nothing
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RadixBasicContainer.prototype, "_id", {
        get: function () {
            return this.hid.toString();
        },
        set: function (_id) {
            // Do nothing
        },
        enumerable: true,
        configurable: true
    });
    RadixBasicContainer.prototype.getSize = function () {
        return this.toByteArray().length;
    };
    return RadixBasicContainer;
}());
exports.default = RadixBasicContainer;
