// radixdlt/build/main/modules/atom/RadixPayloadAtom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixECIES_1 = tslib_1.__importDefault(require("../crypto/RadixECIES"));
var RadixAtomModel_1 = require("../RadixAtomModel");
var elliptic_1 = tslib_1.__importDefault(require("elliptic"));
var ec = new elliptic_1.default.ec('secp256k1');
var RadixPayloadAtom = /** @class */ (function (_super) {
    tslib_1.__extends(RadixPayloadAtom, _super);
    function RadixPayloadAtom(json) {
        var _this = _super.call(this, json) || this;
        _this.serializationProperties.push('encrypted');
        _this.serializationProperties.push('encryptor');
        return _this;
    }
    RadixPayloadAtom.prototype.getDecryptedPayload = function (keyPair) {
        if (this.encrypted && this.encryptor) {
            var rawPayload = this.encryptor.decrypt(this.encrypted, keyPair);
            return rawPayload.toString();
        }
        else if (this.encrypted) {
            var payload = this.encrypted.data.toString();
            return payload;
        }
        throw new Error('No payload');
    };
    RadixPayloadAtom.prototype.addEncryptedPayload = function (payload, recipients) {
        var e_1, _a;
        // Generate key pair
        var ephemeral = ec.genKeyPair();
        // Encrypt key with receivers
        var encryptor = new RadixAtomModel_1.RadixEncryptor();
        encryptor.protectors = [];
        try {
            for (var recipients_1 = tslib_1.__values(recipients), recipients_1_1 = recipients_1.next(); !recipients_1_1.done; recipients_1_1 = recipients_1.next()) {
                var recipient = recipients_1_1.value;
                encryptor.protectors.push(new RadixAtomModel_1.RadixBase64(RadixECIES_1.default.encrypt(recipient.getPublic(), Buffer.from(ephemeral.getPrivate('hex'), 'hex'))));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (recipients_1_1 && !recipients_1_1.done && (_a = recipients_1.return)) _a.call(recipients_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.encryptor = encryptor;
        // Encrypt message
        this.encrypted = new RadixAtomModel_1.RadixBase64(RadixECIES_1.default.encrypt(ephemeral.getPublic(), Buffer.from(payload)));
    };
    RadixPayloadAtom.prototype.addUnencryptedPayload = function (payload) {
        // TODO: transaction message payloads are raw strings not json
        this.encrypted = new RadixAtomModel_1.RadixBase64(Buffer.from(payload));
    };
    return RadixPayloadAtom;
}(RadixAtomModel_1.RadixAtom));
exports.default = RadixPayloadAtom;
