// radixdlt/build/main/modules/universe/RadixUniverseConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var universe_development_1 = tslib_1.__importDefault(require("../common/universe_development"));
var universe_highgarden_1 = tslib_1.__importDefault(require("../common/universe_highgarden"));
var universe_alphanet_1 = tslib_1.__importDefault(require("../common/universe_alphanet"));
var universe_alphanet2_1 = tslib_1.__importDefault(require("../common/universe_alphanet2"));
var universe_sunstone_1 = tslib_1.__importDefault(require("../common/universe_sunstone"));
var long_1 = tslib_1.__importDefault(require("long"));
var RadixUniverseConfig = /** @class */ (function () {
    function RadixUniverseConfig(obj) {
        this.port = obj.port;
        this.name = obj.name;
        this.description = obj.description;
        this.type = obj.type;
        this.timestamp = obj.timestamp;
        this.creator = Buffer.from(obj.creator.value, 'base64');
        this.genesis = obj.genesis;
        // this.magic = this.creator.getUID().value
        //     .mul(new BN(31 * 13 * 7 * this.port * this.type))
        //     .mul(new BN(this.timestamp))
        this.magic = obj.magic;
    }
    RadixUniverseConfig.prototype.getMagic = function () {
        return this.magic;
    };
    RadixUniverseConfig.prototype.getMagicByte = function () {
        return long_1.default.fromNumber(this.magic)
            .and(0xff)
            .toNumber();
    };
    RadixUniverseConfig.WINTERFELL = new RadixUniverseConfig(universe_development_1.default);
    RadixUniverseConfig.WINTERFELL_LOCAL = new RadixUniverseConfig(universe_development_1.default);
    RadixUniverseConfig.SUNSTONE = new RadixUniverseConfig(universe_sunstone_1.default);
    RadixUniverseConfig.HIGHGARDEN = new RadixUniverseConfig(universe_highgarden_1.default);
    RadixUniverseConfig.ALPHANET = new RadixUniverseConfig(universe_alphanet_1.default);
    RadixUniverseConfig.ALPHANET2 = new RadixUniverseConfig(universe_alphanet2_1.default);
    return RadixUniverseConfig;
}());
exports.default = RadixUniverseConfig;
