// radixdlt/build/main/modules/universe/RadixNodeDiscoveryFromSeed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var axios_1 = tslib_1.__importDefault(require("axios"));
var RadixAtomModel_1 = require("../RadixAtomModel");
/**
 * Radix node discovery from a seed node
 */
var RadixNodeDiscoveryFromSeed = /** @class */ (function () {
    /**
     * Creates an instance of radix node discovery from seed.
     * @param bootstrapNode Full address to the rpc endpoint of a Radix node in the universe
     */
    function RadixNodeDiscoveryFromSeed(bootstrapNode) {
        this.bootstrapNode = bootstrapNode;
    }
    RadixNodeDiscoveryFromSeed.prototype.loadNodes = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var getPeersRequestData, nodeListResponse;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        getPeersRequestData = {
                            id: 0,
                            method: 'Network.getLivePeers',
                            params: []
                        };
                        return [4 /*yield*/, axios_1.default.post(this.bootstrapNode, getPeersRequestData)];
                    case 1:
                        nodeListResponse = _a.sent();
                        return [2 /*return*/, RadixAtomModel_1.RadixSerializer.fromJson(nodeListResponse.data.result)];
                }
            });
        });
    };
    return RadixNodeDiscoveryFromSeed;
}());
exports.default = RadixNodeDiscoveryFromSeed;
