// radixdlt/build/module/modules/token/RadixTokenClass.js
import { Decimal } from 'decimal.js';
import { RadixAtom } from '../RadixAtomModel';
export var RadixTokenFlags;
(function (RadixTokenFlags) {
    RadixTokenFlags[RadixTokenFlags["TOKEN_TRADEABLE"] = 1] = "TOKEN_TRADEABLE";
    RadixTokenFlags[RadixTokenFlags["TOKEN_SPENDABLE"] = 2] = "TOKEN_SPENDABLE";
    RadixTokenFlags[RadixTokenFlags["TOKEN_REDEEMABLE"] = 4] = "TOKEN_REDEEMABLE";
    RadixTokenFlags[RadixTokenFlags["TOKEN_RECOVERABLE"] = 8] = "TOKEN_RECOVERABLE";
    RadixTokenFlags[RadixTokenFlags["TOKEN_CHARGABLE"] = 16] = "TOKEN_CHARGABLE";
    RadixTokenFlags[RadixTokenFlags["TOKEN_DIVIDENDS"] = 1024] = "TOKEN_DIVIDENDS";
    RadixTokenFlags[RadixTokenFlags["TOKEN_SYSTEM"] = 4096] = "TOKEN_SYSTEM";
    RadixTokenFlags[RadixTokenFlags["TOKEN_STABILIZED"] = 16384] = "TOKEN_STABILIZED"; // Token is stabilised
})(RadixTokenFlags || (RadixTokenFlags = {}));
export default class RadixTokenClass extends RadixAtom {
    constructor(json) {
        super(json);
        this.serializationProperties.push('type');
        this.serializationProperties.push('iso');
        this.serializationProperties.push('description');
        this.serializationProperties.push('classification');
        this.serializationProperties.push('icon');
        this.serializationProperties.push('sub_units');
        this.serializationProperties.push('maximum_units');
        this.serializationProperties.push('settings');
    }
    /**
     * Convert actual decimal token amount to integer subunits stored on the ledger
     * @param amount
     * @returns subunits
     */
    toSubunits(amount) {
        let x = new Decimal(amount);
        let y = new Decimal(this.sub_units);
        return x
            .times(y)
            .truncated()
            .toNumber();
    }
    /**
     * Convert subunits token amount to actual decimal token amount
     * @param amount
     * @returns token units
     */
    toTokenUnits(amount) {
        let x = new Decimal(amount);
        let y = new Decimal(this.sub_units);
        return x.dividedBy(y).toNumber();
    }
}
RadixTokenClass.SERIALIZER = 62583504;
