// radixdlt/build/module/modules/token/RadixTokenManager.js
import { RadixTokenClass, RadixSerializer } from '../RadixAtomModel';
import { radixUniverse } from '../..';
/**
 * Tokens' information manager.
 */
export class RadixTokenManager {
    constructor() {
        this.tokens = {};
    }
    initialize() {
        for (const atom of radixUniverse.universeConfig.genesis) {
            if (atom.serializer === RadixTokenClass.SERIALIZER) {
                this.addOrUpdateToken(RadixSerializer.fromJson(atom));
            }
        }
    }
    /**
     * Add or update a token, indexed by its ID.
     *
     * @param {RadixTokenClass} token
     * @memberof RadixToken
     */
    addOrUpdateToken(token) {
        this.tokens[token.id.toString()] = token;
    }
    /**
     * Get a token by ID, if it doesn't exists it searches for it in the Ledger.
     *
     * @param {string} id
     * @returns {RadixTokenClass}
     * @memberof RadixToken
     */
    getTokenByID(id) {
        const token = this.tokens[id];
        if (!token) {
            // Search token by id in Ledger?
        }
        return token;
    }
    /**
     * Get a token by ISO, if it doesn't exists it searches for it in the Ledger.
     *
     * @param {string} iso
     * @returns {RadixTokenClass}
     * @memberof RadixToken
     */
    getTokenByISO(iso) {
        for (const token of Object.values(this.tokens)) {
            if (token.iso === iso) {
                return token;
            }
        }
        // Search token by iso in Ledger?
        return null;
    }
    /**
     * Return a list of the current tokens in the manager.
     *
     * @returns
     * @memberof RadixToken
     */
    getCurrentTokens() {
        return this.tokens;
    }
}
export let radixTokenManager = new RadixTokenManager();
