// radixdlt/build/module/modules/atom/RadixPayloadAtom.js
import RadixECIES from '../crypto/RadixECIES';
import { RadixAtom, RadixBase64, RadixEncryptor } from '../RadixAtomModel';
import EC from 'elliptic';
const ec = new EC.ec('secp256k1');
export default class RadixPayloadAtom extends RadixAtom {
    constructor(json) {
        super(json);
        this.serializationProperties.push('encrypted');
        this.serializationProperties.push('encryptor');
    }
    getDecryptedPayload(keyPair) {
        if (this.encrypted && this.encryptor) {
            const rawPayload = this.encryptor.decrypt(this.encrypted, keyPair);
            return rawPayload.toString();
        }
        else if (this.encrypted) {
            const payload = this.encrypted.data.toString();
            return payload;
        }
        throw new Error('No payload');
    }
    addEncryptedPayload(payload, recipients) {
        // Generate key pair
        let ephemeral = ec.genKeyPair();
        // Encrypt key with receivers
        let encryptor = new RadixEncryptor();
        encryptor.protectors = [];
        for (let recipient of recipients) {
            encryptor.protectors.push(new RadixBase64(RadixECIES.encrypt(recipient.getPublic(), Buffer.from(ephemeral.getPrivate('hex'), 'hex'))));
        }
        this.encryptor = encryptor;
        // Encrypt message
        this.encrypted = new RadixBase64(RadixECIES.encrypt(ephemeral.getPublic(), Buffer.from(payload)));
    }
    addUnencryptedPayload(payload) {
        // TODO: transaction message payloads are raw strings not json
        this.encrypted = new RadixBase64(Buffer.from(payload));
    }
}
