// radixdlt/build/module/modules/atom/RadixBasicContainer.js
import { TSMap } from 'typescript-map';
import RadixUtil from '../common/RadixUtil';
import { RadixSerializer, RadixEUID, DataTypes } from '../RadixAtomModel';
export default class RadixBasicContainer {
    constructor(json) {
        this.version = 100;
        this.serializationProperties = new Array();
        if (json) {
            for (let key in json) {
                if (key == 'constructor' || key == 'serializationProperties') {
                    continue;
                }
                this[key] = json[key];
            }
        }
        // this.serializationProperties.push('serializer')
        this.serializationProperties.push('version');
    }
    get serializer() {
        return this.constructor['SERIALIZER'];
    }
    set serializer(serializer) {
        // Do nothing
    }
    toJson() {
        let output = { serializer: 0 };
        for (let key in this) {
            let serialized = RadixSerializer.toJson(this[key]);
            if (serialized) {
                output[key] = serialized;
            }
        }
        output.serializer = this.constructor['SERIALIZER'];
        return output;
    }
    toByteArray() {
        // Generic object
        let type = DataTypes.OBJECT;
        let length = 0;
        // Serialize all properties
        // Build a map sorted by property name
        let map = new TSMap();
        for (let key of this.serializationProperties) {
            if (!(key in this)) {
                continue;
            }
            let serializedValue = RadixSerializer.toByteArray(this[key]);
            length += key.length + 1 + serializedValue.length;
            map.sortedSet(key, serializedValue);
        }
        // Write everything to the buffer
        let output = Buffer.alloc(length + 5);
        output.writeUInt8(type, 0);
        output.writeUInt32BE(length, 1);
        let position = 5;
        for (let key of map.keys()) {
            let value = map.get(key);
            output.writeUInt8(key.length, position);
            position++;
            output.write(key, position);
            position += key.length;
            value.copy(output, position);
            position += value.length;
        }
        return output;
    }
    getHash() {
        let serialized = this.toByteArray();
        return RadixUtil.hash(serialized);
    }
    getHID() {
        let hash = this.getHash();
        return new RadixEUID(hash.slice(0, 12));
    }
    get hid() {
        return this.getHID();
    }
    set hid(hid) {
        // Do nothing
    }
    get _id() {
        return this.hid.toString();
    }
    set _id(_id) {
        // Do nothing
    }
    getSize() {
        return this.toByteArray().length;
    }
}
