// radixdlt/build/module/modules/common/RadixEUID.js
import RadixUtil from './RadixUtil';
import { DataTypes } from '../RadixAtomModel';
import BN from 'bn.js';
export default class RadixEUID {
    constructor(value) {
        if (Array.isArray(value)) {
            this.value = RadixUtil.bigIntFromByteArray(Buffer.from(value));
        }
        else if (Buffer.isBuffer(value)) {
            this.value = RadixUtil.bigIntFromByteArray(value);
        }
        else {
            this.value = new BN(value);
        }
    }
    static fromJson(data) {
        return new RadixEUID(data.value);
    }
    equals(euid) {
        return this.value.eq(euid.value);
    }
    toJson() {
        return {
            serializer: RadixEUID.SERIALIZER,
            value: this.value.toString()
        };
    }
    toByteArray() {
        let type = DataTypes.EUID;
        let data = RadixUtil.byteArrayFromBigInt(this.value);
        let length = data.length;
        let output = Buffer.alloc(length + 5);
        output.writeUInt8(type, 0);
        output.writeUInt32BE(length, 1);
        data.copy(output, 5);
        return output;
    }
    toString() {
        return this.value.toString();
    }
}
RadixEUID.SERIALIZER = 'EUID';
