// radixdlt/build/module/modules/cache/RadixNEDBAtomCache.js
import Datastore from 'nedb';
import { RadixSerializer } from '../RadixAtomModel';
import { logger } from '../common/RadixLogger';
export default class RadixNEDBAtomCache {
    /**
     * Creates an instance of radix nedbatom cache.
     * @param filename File path on disk in Node, path in localStorage in the browser
     */
    constructor(filename) {
        this.storeAtom = (atom) => {
            return this.notExists({ _id: atom._id })
                .then(() => {
                // logger.info('Atom doesnt exist, storing ', atom._id, atom)
                // Add particle ids?
                // Serialize
                const serializedAtom = atom.toJson();
                serializedAtom['_id'] = atom._id;
                // logger.info(serializedAtom)
                // Store
                return this.insert(serializedAtom);
            })
                .then((newDoc) => {
                // Success
                // logger.info('Atom stored in DB', newDoc)
                return atom;
            })
                .catch(error => {
                logger.trace('Atom already in DB');
            });
        };
        this.getAtoms = (keyPair) => {
            // Find
            let query = {};
            // Filter by destination
            if (keyPair) {
                const destination = keyPair.getUID().toJson();
                query = { destinations: destination };
            }
            // logger.info(query)
            return this.find(query).then(async (atoms) => {
                // logger.info(atoms)
                // Deserialize
                const deserialized = [];
                for (const atom of atoms) {
                    deserialized.push(await this.asyncDeserialize(atom));
                }
                return deserialized;
            });
        };
        // Promise wrappers for nedb
        this.findOne = (opt) => {
            return new Promise((resolve, reject) => {
                this.db.findOne(opt, (error, doc) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.notExists = (opt) => {
            return new Promise((resolve, reject) => {
                this.db.findOne(opt, (error, doc) => {
                    if (error) {
                        reject(error);
                    }
                    else if (!doc) {
                        resolve(true);
                    }
                    reject('Atom already in db');
                });
            });
        };
        this.find = (opt) => {
            return new Promise((resolve, reject) => {
                this.db.find(opt, (error, doc) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.insert = (opt) => {
            return new Promise((resolve, reject) => {
                this.db.insert(opt, (error, doc) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.remove = (opt) => {
            return new Promise((resolve, reject) => {
                this.db.remove(opt, (error, doc) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.db = new Datastore({
            filename,
            autoload: true,
        });
    }
    /**
     * Clears all atoms from the cache
     */
    reset() {
        this.db.remove({}, { multi: true }, (error, numRemoved) => {
            // Do nothing
        });
    }
    asyncDeserialize(atom) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve(RadixSerializer.fromJson(atom));
            }, 0);
        });
    }
    deleteAtom(atom) {
        return this.remove({ _id: atom._id });
    }
}
