// radixdlt/build/main/modules/messaging/RadixMessagingAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var RadixAtomModel_1 = require("../RadixAtomModel");
var RadixMessagingAccountSystem = /** @class */ (function () {
    function RadixMessagingAccountSystem(keyPair) {
        this.keyPair = keyPair;
        this.name = 'RADIX-MESSAGING';
        this.messageSubject = new rxjs_1.Subject();
        this.chats = new typescript_map_1.TSMap();
        this.messages = new typescript_map_1.TSMap();
    }
    RadixMessagingAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var atom;
            return tslib_1.__generator(this, function (_a) {
                atom = atomUpdate.atom;
                if (atom.serializer !== RadixAtomModel_1.RadixApplicationPayloadAtom.SERIALIZER || atom.applicationId !== 'radix-messaging') {
                    return [2 /*return*/];
                }
                if (atomUpdate.action === 'STORE') {
                    this.processStoreAtom(atom);
                }
                else if (atomUpdate.action === 'DELETE') {
                    this.processDeleteAtom(atom);
                }
                return [2 /*return*/];
            });
        });
    };
    RadixMessagingAccountSystem.prototype.startNewChat = function (to) {
        // Create new chat
        var chatId = to.getAddress();
        if (this.chats.has(chatId)) {
            return;
        }
        var chatDescription = {
            address: to.getAddress(),
            chat_id: chatId,
            title: chatId,
            last_message_timestamp: Date.now(),
            messages: new typescript_map_1.TSMap()
        };
        // Add at the top
        this.chats.set(chatId, chatDescription);
    };
    RadixMessagingAccountSystem.prototype.processStoreAtom = function (atom) {
        var hid = atom.hid.toString();
        // Skip existing atoms
        if (this.messages.has(hid)) {
            return;
        }
        if (atom.payload === null) {
            return;
        }
        // Format message
        var payload = JSON.parse(atom.payload);
        // TODO: check owner
        var to = RadixAtomModel_1.RadixKeyPair.fromAddress(payload.to);
        var from = RadixAtomModel_1.RadixKeyPair.fromAddress(payload.from);
        // Chat id
        var address = null;
        var isMine = false;
        if (from.equals(this.keyPair)) {
            address = to;
            isMine = true;
        }
        else if (to.equals(this.keyPair)) {
            address = from;
        }
        if (address === null) {
            throw new Error('Error processing a radix-message atom: neither of addresses is owned by this account');
        }
        var chatId = address.toString();
        var message = {
            hid: hid,
            chat_id: chatId,
            to: to,
            from: from,
            content: payload.content,
            timestamp: atom.timestamps.default,
            is_mine: isMine
        };
        // Find existing chat
        // Otherwise create new chat
        if (!this.chats.has(chatId)) {
            var newChatDescription = {
                address: address.getAddress(),
                chat_id: chatId,
                title: chatId,
                last_message_timestamp: message.timestamp,
                messages: new typescript_map_1.TSMap(),
            };
            this.chats.set(chatId, newChatDescription);
        }
        var chatDescription = this.chats.get(chatId);
        if (message.timestamp > chatDescription.last_message_timestamp) {
            chatDescription.last_message_timestamp = message.timestamp;
        }
        chatDescription.messages.set(hid, message);
        // Move chat to the top
        this.chats.delete(chatId);
        this.chats.set(chatId, chatDescription);
        this.messages.set(hid, message);
        var messageUpdate = {
            action: 'STORE',
            hid: hid,
            message: message,
        };
        this.messageSubject.next(messageUpdate);
    };
    RadixMessagingAccountSystem.prototype.processDeleteAtom = function (atom) {
        var hid = atom.hid.toString();
        // Skip nonexisting atoms
        if (!this.messages.has(hid)) {
            return;
        }
        var message = this.messages.get(hid);
        this.chats.get(message.chat_id).messages.delete(hid);
        this.messages.delete(hid);
        var messageUpdate = {
            action: 'DELETE',
            hid: hid,
            message: message,
        };
        this.messageSubject.next(messageUpdate);
    };
    RadixMessagingAccountSystem.prototype.getAllMessages = function () {
        var _this = this;
        return rxjs_1.Observable.create(function (observer) {
            var e_1, _a;
            try {
                // Send all old transactions
                for (var _b = tslib_1.__values(_this.messages.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var message = _c.value;
                    var messageUpdate = {
                        action: 'STORE',
                        hid: message.hid,
                        message: message
                    };
                    observer.next(messageUpdate);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            // Subscribe for new ones
            _this.messageSubject.subscribe(observer);
        });
    };
    return RadixMessagingAccountSystem;
}());
exports.default = RadixMessagingAccountSystem;
