// radixdlt/build/module/modules/account/RadixAccount.js
import { BehaviorSubject } from 'rxjs';
import { TSMap } from 'typescript-map';
import RadixKeyPair from '../wallet/RadixKeyPair';
import RadixDataAccountSystem from './RadixDataAccountSystem';
import { radixUniverse } from '../universe/RadixUniverse';
import { RadixTransferAccountSystem, RadixMessagingAccountSystem, RadixDecryptionAccountSystem, RadixCacheAccountSystem } from '../..';
import { logger } from '../common/RadixLogger';
export default class RadixAccount {
    /**
     * Creates an instance of radix account.
     * @param keyPair Public key of the account
     * @param [plain] If set to false, will not create default account systems.
     * Use this for accounts that will not be connected to the network
     */
    constructor(keyPair, plain = false) {
        this.keyPair = keyPair;
        this.accountSystems = new TSMap();
        this.connectionStatus = new BehaviorSubject('STARTING');
        this.openNodeConnection = async () => {
            this.connectionStatus.next('CONNECTING');
            try {
                this.nodeConnection = await radixUniverse.getNodeConnection(this.keyPair.getShard());
                this.connectionStatus.next('CONNECTED');
                this.nodeConnection.on('closed', this._onConnectionClosed);
                // Subscribe to events
                this.atomSubscription = this.nodeConnection.subscribe(this.keyPair.toString());
                this.atomSubscription.subscribe({
                    next: this._onAtomReceived,
                    error: error => logger.error('Subscription error:', error)
                });
            }
            catch (error) {
                logger.error(error);
                setTimeout(this._onConnectionClosed, 1000);
            }
        };
        this._onAtomReceived = async (atomUpdate) => {
            for (const system of this.accountSystems.values()) {
                await system.processAtomUpdate(atomUpdate);
            }
        };
        this._onConnectionClosed = () => {
            // Get a new one
            this.openNodeConnection();
        };
        if (!plain) {
            this.cacheSystem = new RadixCacheAccountSystem(keyPair);
            this.addAccountSystem(this.cacheSystem);
            this.decryptionSystem = new RadixDecryptionAccountSystem();
            this.addAccountSystem(this.decryptionSystem);
            this.transferSystem = new RadixTransferAccountSystem(keyPair);
            this.addAccountSystem(this.transferSystem);
            this.dataSystem = new RadixDataAccountSystem(keyPair);
            this.addAccountSystem(this.dataSystem);
            this.messagingSystem = new RadixMessagingAccountSystem(keyPair);
            this.addAccountSystem(this.messagingSystem);
        }
    }
    /**
     * Create an instance of radix account from an address
     * @param address string address
     * @param [plain] If set to false, will not create default account systems.
     * Use this for accounts that will not be connected to the network
     * @returns
     */
    static fromAddress(address, plain = false) {
        return new RadixAccount(RadixKeyPair.fromAddress(address), plain);
    }
    enableDecryption(decryptionProvider) {
        this.decryptionSystem.decryptionProvider = decryptionProvider;
    }
    enableCache(cacheProvider) {
        this.cacheSystem.atomCache = cacheProvider;
        // Load atoms from cache
        return this.cacheSystem.loadAtoms().then((atoms) => {
            for (const atom of atoms) {
                this._onAtomReceived({
                    action: 'STORE',
                    atom,
                });
            }
        });
    }
    getAddress() {
        return this.keyPair.getAddress();
    }
    addAccountSystem(system) {
        if (this.accountSystems.has(system.name)) {
            throw new Error(`System "${system.name}" already exists in account, you can only have one of each system per account`);
        }
        this.accountSystems.set(system.name, system);
        return system;
    }
    removeAccountSystem(name) {
        if (this.accountSystems.has(name)) {
            this.accountSystems.delete(name);
        }
    }
    getSystem(name) {
        if (this.accountSystems.has(name)) {
            return this.accountSystems.get(name);
        }
        throw new Error(`System "${name}" doesn't exist in account`);
    }
}
