// radixdlt/build/module/modules/account/RadixTransferAccountSystem.js
import { Subject, Observable, BehaviorSubject } from 'rxjs';
import { TSMap } from 'typescript-map';
import RadixKeyPair from '../wallet/RadixKeyPair';
import { radixTokenManager } from '../token/RadixTokenManager';
import { radixConfig } from '../common/RadixConfig';
import { RadixTransactionAtom, RadixConsumer, RadixConsumable, RadixEmission, RadixAtomFeeConsumable } from '../RadixAtomModel';
export default class RadixTransferAccountSystem {
    constructor(keyPair) {
        this.keyPair = keyPair;
        this.name = 'TRANSFER';
        this.transactions = new TSMap();
        this.balance = {};
        this.transactionSubject = new Subject();
        this.unspentConsumables = new TSMap();
        this.spentConsumables = new TSMap();
        // Add default radix token to balance
        this.balance[radixTokenManager.getTokenByISO(radixConfig.mainTokenISO).id.toString()] = 0;
        this.balanceSubject = new BehaviorSubject(this.balance);
    }
    async processAtomUpdate(atomUpdate) {
        const atom = atomUpdate.atom;
        if (atom.serializer !== RadixTransactionAtom.SERIALIZER) {
            return;
        }
        if (atomUpdate.action === 'STORE') {
            this.processStoreAtom(atom);
        }
        else if (atomUpdate.action === 'DELETE') {
            this.processDeleteAtom(atom);
        }
    }
    processStoreAtom(atom) {
        // Skip existing atoms
        if (this.transactions.has(atom.hid.toString())) {
            return;
        }
        const transactionUpdate = {
            action: 'STORE',
            hid: atom.hid.toString(),
            transaction: {
                hid: atom.hid.toString(),
                balance: {},
                fee: 0,
                participants: {},
                timestamp: atom.timestamps.default,
                message: ''
            }
        };
        const transaction = transactionUpdate.transaction;
        // Get transaction message
        if (typeof atom.payload === 'string') {
            transaction.message = atom.payload;
        }
        // Get transaction details
        for (const particle of atom.particles) {
            const tokenId = particle.asset_id.toString();
            if (!radixTokenManager.getTokenByID(tokenId)) {
                throw new Error('Unsuporeted Token Class');
            }
            let ownedByMe = false;
            for (const owner of particle.owners) {
                if (owner.public.data.equals(this.keyPair.getPublic())) {
                    ownedByMe = true;
                    break;
                }
            }
            const isFee = particle.serializer === RadixAtomFeeConsumable.SERIALIZER;
            if (ownedByMe && !isFee) {
                let quantity = 0;
                if (particle.serializer === RadixConsumer.SERIALIZER) {
                    quantity -= particle.quantity;
                    this.unspentConsumables.delete(particle._id);
                    this.spentConsumables.set(particle._id, particle);
                }
                else if (particle.serializer === RadixConsumable.SERIALIZER ||
                    particle.serializer === RadixEmission.SERIALIZER) {
                    quantity += particle.quantity;
                    if (!this.spentConsumables.has(particle._id)) {
                        this.unspentConsumables.set(particle._id, particle);
                    }
                }
                if (!(tokenId in transaction.balance)) {
                    transaction.balance[tokenId] = 0;
                }
                transaction.balance[tokenId] += quantity;
            }
            else if (!ownedByMe && !isFee) {
                for (const owner of particle.owners) {
                    const keyPair = RadixKeyPair.fromRadixECKeyPair(owner);
                    transaction.participants[keyPair.getAddress()] = keyPair.getAddress();
                }
            }
        }
        this.transactions.set(transactionUpdate.hid, transaction);
        // Update balance
        for (const tokenId in transaction.balance) {
            if (!(tokenId in this.balance)) {
                this.balance[tokenId] = 0;
            }
            this.balance[tokenId] += transaction.balance[tokenId];
        }
        this.balanceSubject.next(this.balance);
        this.transactionSubject.next(transactionUpdate);
    }
    processDeleteAtom(atom) {
        // Skip nonexisting atoms
        if (!this.transactions.has(atom.hid.toString())) {
            return;
        }
        const hid = atom.hid.toString();
        const transaction = this.transactions.get(hid);
        const transactionUpdate = {
            action: 'DELETE',
            hid,
            transaction,
        };
        // Update consumables
        for (const particle of atom.particles) {
            const tokenId = particle.asset_id.toString();
            if (!radixTokenManager.getCurrentTokens()[tokenId]) {
                throw new Error('Unsuporeted Token Class');
            }
            let ownedByMe = false;
            for (const owner of particle.owners) {
                if (owner.public.data.equals(this.keyPair.getPublic())) {
                    ownedByMe = true;
                    break;
                }
            }
            const isFee = particle.serializer === RadixAtomFeeConsumable.SERIALIZER;
            if (ownedByMe && !isFee) {
                if (particle.serializer === RadixConsumer.SERIALIZER) {
                    this.unspentConsumables.set(particle._id, particle);
                    this.spentConsumables.delete(particle._id);
                }
                else if (particle.serializer === RadixConsumable.SERIALIZER ||
                    particle.serializer === RadixEmission.SERIALIZER) {
                    this.unspentConsumables.delete(particle._id);
                }
            }
        }
        // Update balance
        for (const tokenId in transaction.balance) {
            if (!(tokenId in this.balance)) {
                this.balance[tokenId] = 0;
            }
            this.balance[tokenId] -= transaction.balance[tokenId];
        }
        this.balanceSubject.next(this.balance);
        this.transactionSubject.next(transactionUpdate);
    }
    getAllTransactions() {
        return Observable.create((observer) => {
            // Send all old transactions
            for (const transaction of this.transactions.values()) {
                const transactionUpdate = {
                    action: 'STORE',
                    hid: transaction.hid,
                    transaction: transaction
                };
                observer.next(transactionUpdate);
            }
            // Subscribe for new ones
            this.transactionSubject.subscribe(observer);
        });
    }
    getUnspentConsumables() {
        return this.unspentConsumables;
    }
}
