// radixdlt/build/module/modules/account/RadixDataAccountSystem.js
import { Subject, Observable } from 'rxjs';
import { TSMap } from 'typescript-map';
import { filter } from 'rxjs/operators';
import { RadixApplicationPayloadAtom } from '../RadixAtomModel';
export default class RadixDataAccountSystem {
    constructor(keyPair) {
        this.keyPair = keyPair;
        this.name = 'DATA';
        this.applicationDataSubject = new Subject();
        this.applicationData = new TSMap();
    }
    async processAtomUpdate(atomUpdate) {
        if (atomUpdate.atom.serializer !== RadixApplicationPayloadAtom.SERIALIZER) {
            return;
        }
        if (atomUpdate.action === 'STORE') {
            this.processStoreAtom(atomUpdate.atom);
        }
        else if (atomUpdate.action === 'DELETE') {
            this.processDeleteAtom(atomUpdate.atom);
        }
    }
    processStoreAtom(atom) {
        const applicationId = atom.applicationId;
        const hid = atom.hid.toString();
        // Skip existing atoms
        if (this.applicationData.has(applicationId) &&
            this.applicationData.get(applicationId).has(hid)) {
            return;
        }
        const applicationData = {
            hid,
            payload: '',
            timestamp: atom.timestamps.default
        };
        const applicationDataUpdate = {
            action: 'STORE',
            hid,
            applicationId,
            data: applicationData,
        };
        if (atom.payload === null) {
            return;
        }
        applicationData.payload = atom.payload;
        if (!this.applicationData.has(applicationId)) {
            this.applicationData.set(applicationId, new TSMap());
        }
        this.applicationData.get(applicationId).set(hid, applicationData);
        this.applicationDataSubject.next(applicationDataUpdate);
    }
    processDeleteAtom(atom) {
        const applicationId = atom.applicationId;
        const hid = atom.hid.toString();
        // Skip nonexisting atoms
        if (!this.applicationData.has(applicationId) ||
            !this.applicationData.get(applicationId).has(hid)) {
            return;
        }
        const applicationData = this.applicationData.get(applicationId).get(hid);
        const applicationDataUpdate = {
            action: 'DELETE',
            hid,
            applicationId,
            data: applicationData,
        };
        this.applicationData.get(applicationId).delete(hid);
        this.applicationDataSubject.next(applicationDataUpdate);
    }
    getApplicationData(applicationId) {
        return Observable.create((observer) => {
            // Send all old data
            if (this.applicationData.has(applicationId)) {
                for (const applicationData of this.applicationData
                    .get(applicationId)
                    .values()) {
                    const applicationDataUpdate = {
                        action: 'STORE',
                        hid: applicationData.hid,
                        applicationId: applicationId,
                        data: applicationData
                    };
                    observer.next(applicationDataUpdate);
                }
            }
            // Subscribe for new ones
            this.applicationDataSubject
                .pipe(filter(update => {
                return update.applicationId === applicationId;
            }))
                .subscribe(observer);
        });
    }
}
