// radixdlt/build/main/modules/account/RadixDataAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var operators_1 = require("rxjs/operators");
var RadixAtomModel_1 = require("../RadixAtomModel");
var RadixDataAccountSystem = /** @class */ (function () {
    function RadixDataAccountSystem(keyPair) {
        this.keyPair = keyPair;
        this.name = 'DATA';
        this.applicationDataSubject = new rxjs_1.Subject();
        this.applicationData = new typescript_map_1.TSMap();
    }
    RadixDataAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (atomUpdate.atom.serializer !== RadixAtomModel_1.RadixApplicationPayloadAtom.SERIALIZER) {
                    return [2 /*return*/];
                }
                if (atomUpdate.action === 'STORE') {
                    this.processStoreAtom(atomUpdate.atom);
                }
                else if (atomUpdate.action === 'DELETE') {
                    this.processDeleteAtom(atomUpdate.atom);
                }
                return [2 /*return*/];
            });
        });
    };
    RadixDataAccountSystem.prototype.processStoreAtom = function (atom) {
        var applicationId = atom.applicationId;
        var hid = atom.hid.toString();
        // Skip existing atoms
        if (this.applicationData.has(applicationId) &&
            this.applicationData.get(applicationId).has(hid)) {
            return;
        }
        var applicationData = {
            hid: hid,
            payload: '',
            timestamp: atom.timestamps.default
        };
        var applicationDataUpdate = {
            action: 'STORE',
            hid: hid,
            applicationId: applicationId,
            data: applicationData,
        };
        if (atom.payload === null) {
            return;
        }
        applicationData.payload = atom.payload;
        if (!this.applicationData.has(applicationId)) {
            this.applicationData.set(applicationId, new typescript_map_1.TSMap());
        }
        this.applicationData.get(applicationId).set(hid, applicationData);
        this.applicationDataSubject.next(applicationDataUpdate);
    };
    RadixDataAccountSystem.prototype.processDeleteAtom = function (atom) {
        var applicationId = atom.applicationId;
        var hid = atom.hid.toString();
        // Skip nonexisting atoms
        if (!this.applicationData.has(applicationId) ||
            !this.applicationData.get(applicationId).has(hid)) {
            return;
        }
        var applicationData = this.applicationData.get(applicationId).get(hid);
        var applicationDataUpdate = {
            action: 'DELETE',
            hid: hid,
            applicationId: applicationId,
            data: applicationData,
        };
        this.applicationData.get(applicationId).delete(hid);
        this.applicationDataSubject.next(applicationDataUpdate);
    };
    RadixDataAccountSystem.prototype.getApplicationData = function (applicationId) {
        var _this = this;
        return rxjs_1.Observable.create(function (observer) {
            var e_1, _a;
            // Send all old data
            if (_this.applicationData.has(applicationId)) {
                try {
                    for (var _b = tslib_1.__values(_this.applicationData
                        .get(applicationId)
                        .values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var applicationData = _c.value;
                        var applicationDataUpdate = {
                            action: 'STORE',
                            hid: applicationData.hid,
                            applicationId: applicationId,
                            data: applicationData
                        };
                        observer.next(applicationDataUpdate);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            // Subscribe for new ones
            _this.applicationDataSubject
                .pipe(operators_1.filter(function (update) {
                return update.applicationId === applicationId;
            }))
                .subscribe(observer);
        });
    };
    return RadixDataAccountSystem;
}());
exports.default = RadixDataAccountSystem;
